### PR TITLE
Use buffer.Length as the source of truth for ReadDirectoryChanges' nBufferLength

### PR DESCRIPTION
**Pull request by jozkee**

### CODE REVIEW

File: src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs [hunk 1]
```diff
@@ -158,7 +158,7 @@ private unsafe void Monitor(AsyncReadState state)
                 continueExecuting = Interop.Kernel32.ReadDirectoryChangesW(
                     state.DirectoryHandle,
                     state.Buffer, // the buffer is kept pinned for the duration of the sync and async operation by the PreAllocatedOverlapped
-                    _internalBufferSize,
+                    (uint)state.Buffer.Length,
```
Comment (stephentoub) [inline src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs:161]: I assume the problem here was a concurrent change to the InternalBufferSize could lead to corruption?  Is there an issue for this?

Comment (Jozkee) [reply src/libraries/System.IO.FileSystem.Watcher/src/System/IO/FileSystemWatcher.Win32.cs:161]: Correct, this is trying to fix a possible corruption bug; there's not a GitHub issue, though.

**Review by carlossanlop — APPROVED** [review]

**Review by danmoseley — APPROVED** [review]
LGTM. I suppose in theory this could have a test that attempts to force the race, but if we were writing this code fresh, I don't think we'd write such a test, and the fix is self evidently how we should have written it.

**Comment by danmoseley** [comment]
failures so far are the prevalent issue right now... "failed to upload results"

**Review by adamsitnik — APPROVED** [review]
LGTM

**Pull request merged**